File: src-tauri/src/repeat.rs
//! 重复规则（RRULE）解析与下一次出现日期计算。
//!
//! 序列化格式（类 RRULE）：
//!   FREQ=WEEKLY;INTERVAL=2;BYDAY=0,2,4;COUNT=10;UNTIL=2026-12-31T23:59:59
//!
//! 其中 BYDAY 使用数字 0=周日..6=周六，与 chrono 的 `Weekday::num_days_from_sunday()` 一致。

use chrono::{DateTime, Datelike, Local, NaiveDate, TimeZone};

#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
pub struct RepeatRule {
    pub freq: String, // "DAILY" | "WEEKLY" | "MONTHLY" | "YEARLY"
    pub interval: i32,
    pub byweekday: Option<Vec<i32>>, // 0=周日..6=周六
    pub end_date: Option<String>,    // RFC3339 ISO 日期
    pub count: Option<i32>,
}

const VALID_FREQS: &[&str] = &["DAILY", "WEEKLY", "MONTHLY", "YEARLY"];

/// 将 RepeatRule 序列化为 RRULE 格式字符串。
pub fn serialize_rrule(rule: &RepeatRule) -> String {
    let mut parts = vec![
        format!("FREQ={}", rule.freq),
        format!("INTERVAL={}", rule.interval),
    ];
    if let Some(ref days) = rule.byweekday {
        if !days.is_empty() {
            let day_str: Vec<String> = days.iter().map(|d| d.to_string()).collect();
            parts.push(format!("BYDAY={}", day_str.join(",")));
        }
    }
    if let Some(ref end_date) = rule.end_date {
        parts.push(format!("UNTIL={}", end_date));
    }
    if let Some(count) = rule.count {
        if count > 0 {
            parts.push(format!("COUNT={}", count));
        }
    }
    parts.join(";")
}

/// 解析 RRULE 格式字符串为 RepeatRule。
/// 无效输入返回 None。
pub fn parse_rrule(s: &str) -> Option<RepeatRule> {
    let s = s.trim();
    if s.is_empty() {
        return None;
    }

    let mut map: std::collections::HashMap<String, String> = std::collections::HashMap::new();
    for part in s.split(';') {
        let part = part.trim();
        if part.is_empty() {
            continue;
        }
        if let Some(idx) = part.find('=') {
            let key = part[..idx].to_uppercase();
            let val = part[idx + 1..].trim().to_string();
            map.insert(key, val);
        }
    }

    let freq = map.get("FREQ")?.clone();
    if !VALID_FREQS.contains(&freq.as_str()) {
        return None;
    }

    let interval = map
        .get("INTERVAL")
        .and_then(|s| s.parse::<i32>().ok())
        .unwrap_or(1);
    if interval < 1 {
        return None;
    }

    let byweekday = map
        .get("BYDAY")
        .map(|s| {
            s.split(',')
                .filter_map(|d| d.trim().parse::<i32>().ok())
                .filter(|d| *d >= 0 && *d <= 6)
                .collect::<Vec<_>>()
        })
        .filter(|v| !v.is_empty());

    let end_date = map.get("UNTIL").cloned();
    let count = map.get("COUNT").and_then(|s| s.parse::<i32>().ok());

    Some(RepeatRule {
        freq,
        interval,
        byweekday,
        end_date,
        count,
    })
}

/// 返回日期所在周的周日（00:00 的 NaiveDate）。
/// chrono `Weekday::num_days_from_sunday()` 返回 0=周日..6=周六。
fn sunday_of(dt: DateTime<Local>) -> NaiveDate {
    let day = dt.weekday().num_days_from_sunday() as i64;
    dt.date_naive() - chrono::Duration::days(day)
}

/// 根据规则（freq/interval/byweekday）计算从 from 之后的下一个出现日期，
/// 不考虑 endDate / count 限制。
fn compute_next(rule: &RepeatRule, from: DateTime<Local>) -> Option<DateTime<Local>> {
    let interval = rule.interval as i64;

    match rule.freq.as_str() {
        "DAILY" => Some(from + chrono::Duration::days(interval)),

        "WEEKLY" => {
            if let Some(ref days) = rule.byweekday {
                if days.is_empty() {
                    return Some(from + chrono::Duration::weeks(interval));
                }
                let from_sunday = sunday_of(from);
                let limit = 7 * interval + 7;
                for i in 1..=limit {
                    let candidate = from + chrono::Duration::days(i);
                    let cand_day = candidate.weekday().num_days_from_sunday() as i32;
                    if days.contains(&cand_day) {
                        let cand_sunday = sunday_of(candidate);
                        let week_diff = (cand_sunday - from_sunday).num_days() / 7;
                        if week_diff >= 0 && week_diff % interval == 0 {
                            return Some(candidate);
                        }
                    }
                }
                None
            } else {
                Some(from + chrono::Duration::weeks(interval))
            }
        }

        "MONTHLY" => {
            let original_day = from.day();
            let mut year = from.year();
            let mut month = from.month() as i32 + interval as i32;
            while month > 12 {
                month -= 12;
                year += 1;
            }
            // 尝试创建目标日期，若该月无此日则回退到月末
            let target_naive =
                NaiveDate::from_ymd_opt(year, month as u32, original_day).or_else(|| {
                    // 目标月最后一天 = 下月1号 - 1天
                    let next_first = if month == 12 {
                        NaiveDate::from_ymd_opt(year + 1, 1, 1)
                    } else {
                        NaiveDate::from_ymd_opt(year, month as u32 + 1, 1)
                    }?;
                    Some(next_first - chrono::Duration::days(1))
                })?;
            let naive_dt = target_naive.and_time(from.naive_local().time());
            Local.from_local_datetime(&naive_dt).single()
        }

        "YEARLY" => {
            let original_month = from.month();
            let original_day = from.day();
            let target_year = from.year() + interval as i32;
            // 尝试创建目标日期，若无效（如闰年2月29日）则回退到目标月月末
            let target_naive = NaiveDate::from_ymd_opt(target_year, original_month, original_day)
                .or_else(|| {
                let next_first = if original_month == 12 {
                    NaiveDate::from_ymd_opt(target_year + 1, 1, 1)
                } else {
                    NaiveDate::from_ymd_opt(target_year, original_month + 1, 1)
                }?;
                Some(next_first - chrono::Duration::days(1))
            })?;
            let naive_dt = target_naive.and_time(from.naive_local().time());
            Local.from_local_datetime(&naive_dt).single()
        }

        _ => None,
    }
}

/// 根据规则计算从 from 之后的下一个出现日期。
/// 返回 None 表示规则已到期（endDate/count 到达）或无法计算。
pub fn next_occurrence(rule: &RepeatRule, from: DateTime<Local>) -> Option<DateTime<Local>> {
    // count 已耗尽：count<=0 表示不再生成新出现
    if let Some(count) = rule.count {
        if count <= 0 {
            return None;
        }
    }

    let next = compute_next(rule, from)?;

    // endDate 到期检查
    if let Some(ref end_date) = rule.end_date {
        if let Ok(end_dt) = chrono::DateTime::parse_from_rfc3339(end_date) {
            let end_local = end_dt.with_timezone(&Local);
            if next > end_local {
                return None;
            }
        }
    }

    Some(next)
}

#[cfg(test)]
mod tests {
    use super::*;
    use chrono::TimeZone;

    #[test]
    fn test_serialize_daily() {
        let rule = RepeatRule {
            freq: "DAILY".to_string(),
            interval: 1,
            byweekday: None,
            end_date: None,
            count: None,
        };
        assert_eq!(serialize_rrule(&rule), "FREQ=DAILY;INTERVAL=1");
    }

    #[test]
    fn test_serialize_weekly_with_byday() {
        let rule = RepeatRule {
            freq: "WEEKLY".to_string(),
            interval: 2,
            byweekday: Some(vec![0, 2, 4]),
            end_date: None,
            count: Some(10),
        };
        let result = serialize_rrule(&rule);
        assert!(result.contains("FREQ=WEEKLY"));
        assert!(result.contains("INTERVAL=2"));
        assert!(result.contains("BYDAY=0,2,4"));
        assert!(result.contains("COUNT=10"));
    }

    #[test]
    fn test_parse_valid_rrule() {
        let rule = parse_rrule("FREQ=DAILY;INTERVAL=2").unwrap();
        assert_eq!(rule.freq, "DAILY");
        assert_eq!(rule.interval, 2);
        assert!(rule.byweekday.is_none());
        assert!(rule.end_date.is_none());
        assert!(rule.count.is_none());
    }

    #[test]
    fn test_parse_empty_string() {
        assert!(parse_rrule("").is_none());
        assert!(parse_rrule("   ").is_none());
    }

    #[test]
    fn test_parse_invalid_freq() {
        assert!(parse_rrule("FREQ=HOURLY;INTERVAL=1").is_none());
        assert!(parse_rrule("FREQ=invalid;INTERVAL=1").is_none());
    }

    #[test]
    fn test_parse_invalid_interval() {
        assert!(parse_rrule("FREQ=DAILY;INTERVAL=0").is_none());
        assert!(parse_rrule("FREQ=DAILY;INTERVAL=-1").is_none());
    }

    #[test]
    fn test_parse_default_interval() {
        let rule = parse_rrule("FREQ=WEEKLY").unwrap();
        assert_eq!(rule.interval, 1);
    }

    #[test]
    fn test_parse_with_until() {
        let rule = parse_rrule("FREQ=DAILY;INTERVAL=1;UNTIL=2026-12-31T23:59:59").unwrap();
        assert_eq!(rule.end_date, Some("2026-12-31T23:59:59".to_string()));
    }

    #[test]
    fn test_parse_byday_filter() {
        // 合法日期 0-6 被保留，非法日期（7、8）被过滤
        let rule = parse_rrule("FREQ=WEEKLY;INTERVAL=1;BYDAY=0,2,7,8").unwrap();
        assert_eq!(rule.byweekday, Some(vec![0, 2]));

        // 全部非法 → 过滤为空 → None
        let rule = parse_rrule("FREQ=WEEKLY;INTERVAL=1;BYDAY=7,8").unwrap();
        assert!(rule.byweekday.is_none());
    }

    #[test]
    fn test_next_occurrence_daily() {
        let rule = RepeatRule {
            freq: "DAILY".to_string(),
            interval: 1,
            byweekday: None,
            end_date: None,
            count: None,
        };
        let from = Local.with_ymd_and_hms(2026, 1, 1, 9, 0, 0).unwrap();
        let next = next_occurrence(&rule, from).unwrap();
        assert_eq!(next, from + chrono::Duration::days(1));
    }

    #[test]
    fn test_next_occurrence_daily_interval_3() {
        let rule = RepeatRule {
            freq: "DAILY".to_string(),
            interval: 3,
            byweekday: None,
            end_date: None,
            count: None,
        };
        let from = Local.with_ymd_and_hms(2026, 1, 1, 9, 0, 0).unwrap();
        let next = next_occurrence(&rule, from).unwrap();
        assert_eq!(next, from + chrono::Duration::days(3));
    }

    #[test]
    fn test_next_occurrence_weekly_no_byday() {
        let rule = RepeatRule {
            freq: "WEEKLY".to_string(),
            interval: 1,
            byweekday: None,
            end_date: None,
            count: None,
        };
        let from = Local.with_ymd_and_hms(2026, 1, 1, 9, 0, 0).unwrap();
        let next = next_occurrence(&rule, from).unwrap();
        assert_eq!(next, from + chrono::Duration::weeks(1));
    }

    #[test]
    fn test_next_occurrence_count_exhausted() {
        let rule = RepeatRule {
            freq: "DAILY".to_string(),
            interval: 1,
            byweekday: None,
            end_date: None,
            count: Some(0),
        };
        let from = Local.with_ymd_and_hms(2026, 1, 1, 9, 0, 0).unwrap();
        assert!(next_occurrence(&rule, from).is_none());

        // 负数 count 同样视为已耗尽
        let rule_neg = RepeatRule {
            freq: "DAILY".to_string(),
            interval: 1,
            byweekday: None,
            end_date: None,
            count: Some(-1),
        };
        assert!(next_occurrence(&rule_neg, from).is_none());
    }

    #[test]
    fn test_next_occurrence_end_date_reached() {
        let rule = RepeatRule {
            freq: "DAILY".to_string(),
            interval: 1,
            byweekday: None,
            end_date: Some("2026-01-01T23:59:59Z".to_string()),
            count: None,
        };
        // from = 2026-01-01 09:00，下一次出现 = 2026-01-02 09:00，已超过 end_date
        let from = Local.with_ymd_and_hms(2026, 1, 1, 9, 0, 0).unwrap();
        assert!(next_occurrence(&rule, from).is_none());
    }

    #[test]
    fn test_next_occurrence_monthly() {
        let rule = RepeatRule {
            freq: "MONTHLY".to_string(),
            interval: 1,
            byweekday: None,
            end_date: None,
            count: None,
        };
        let from = Local.with_ymd_and_hms(2026, 1, 15, 9, 0, 0).unwrap();
        let next = next_occurrence(&rule, from).unwrap();
        assert_eq!(next, Local.with_ymd_and_hms(2026, 2, 15, 9, 0, 0).unwrap());
    }

    #[test]
    fn test_next_occurrence_yearly() {
        let rule = RepeatRule {
            freq: "YEARLY".to_string(),
            interval: 1,
            byweekday: None,
            end_date: None,
            count: None,
        };
        let from = Local.with_ymd_and_hms(2026, 6, 15, 9, 0, 0).unwrap();
        let next = next_occurrence(&rule, from).unwrap();
        assert_eq!(next, Local.with_ymd_and_hms(2027, 6, 15, 9, 0, 0).unwrap());
    }
}

File: src/types/repeat.ts
/**
 * 重复规则类型与工具函数。
 *
 * 序列化格式（类 RRULE）：
 *   FREQ=WEEKLY;INTERVAL=2;BYDAY=0,2,4;COUNT=10;UNTIL=2026-12-31T23:59:59
 *
 * 其中 BYDAY 使用数字 0=周日..6=周六，与 JS Date#getDay() 一致。
 */

export type RepeatFrequency = 'DAILY' | 'WEEKLY' | 'MONTHLY' | 'YEARLY'

export interface RepeatRule {
  freq: RepeatFrequency
  interval: number // 每 N 天/周/月/年
  byweekday?: number[] // WEEKLY 时生效，0=周日..6=周六
  endDate?: string // ISO 日期，可选
  count?: number // 重复次数，可选（与 endDate 互斥）
}

// 星期中文名映射（0=周日..6=周六）
const WEEKDAY_NAMES = ['日', '一', '二', '三', '四', '五', '六']

const VALID_FREQS: RepeatFrequency[] = ['DAILY', 'WEEKLY', 'MONTHLY', 'YEARLY']

const MS_PER_WEEK = 7 * 24 * 60 * 60 * 1000

/**
 * 解析 RRULE 格式字符串为 RepeatRule 对象。
 * 支持格式：FREQ=WEEKLY;INTERVAL=2;BYDAY=0,2,4;COUNT=10;UNTIL=2026-12-31
 * 无效输入返回 null。
 */
export function parseRepeatRule(rule: string | null | undefined): RepeatRule | null {
  if (!rule || typeof rule !== 'string' || rule.trim() === '') return null

  const map: Record<string, string> = {}
  const parts = rule
    .split(';')
    .map((p) => p.trim())
    .filter(Boolean)
  for (const part of parts) {
    const idx = part.indexOf('=')
    if (idx === -1) continue
    const key = part.slice(0, idx).toUpperCase().trim()
    const val = part.slice(idx + 1).trim()
    map[key] = val
  }

  const freq = map['FREQ'] as RepeatFrequency | undefined
  if (!freq || !VALID_FREQS.includes(freq)) return null

  const interval = map['INTERVAL'] ? parseInt(map['INTERVAL'], 10) : 1
  if (isNaN(interval) || interval < 1) return null

  const result: RepeatRule = { freq, interval }

  if (map['BYDAY']) {
    const days = map['BYDAY']
      .split(',')
      .map((d) => parseInt(d.trim(), 10))
      .filter((d) => !isNaN(d) && d >= 0 && d <= 6)
    if (days.length > 0) {
      result.byweekday = days
    }
  }

  if (map['UNTIL']) {
    result.endDate = map['UNTIL']
  }

  if (map['COUNT']) {
    const count = parseInt(map['COUNT'], 10)
    if (!isNaN(count)) {
      result.count = count
    }
  }

  return result
}

/**
 * 将 RepeatRule 序列化为 RRULE 格式字符串。
 */
export function serializeRepeatRule(rule: RepeatRule): string {
  const parts: string[] = [`FREQ=${rule.freq}`, `INTERVAL=${rule.interval}`]
  if (rule.byweekday && rule.byweekday.length > 0) {
    parts.push(`BYDAY=${rule.byweekday.join(',')}`)
  }
  if (rule.endDate) {
    parts.push(`UNTIL=${rule.endDate}`)
  }
  if (rule.count !== undefined && rule.count > 0) {
    parts.push(`COUNT=${rule.count}`)
  }
  return parts.join(';')
}

/**
 * 返回从 from 所在周的周日 00:00 的 Date。
 * JS Date#getDay() 中 0=周日..6=周六。
 */
function startOfWeek(date: Date): Date {
  const d = new Date(date)
  const day = d.getDay() // 0=周日
  d.setDate(d.getDate() - day)
  d.setHours(0, 0, 0, 0)
  return d
}

/**
 * 根据规则（freq/interval/byweekday）计算从 from 之后的下一个出现日期，
 * 不考虑 endDate / count 限制。
 */
function computeNextDate(rule: RepeatRule, from: Date): Date | null {
  const { freq, interval } = rule

  switch (freq) {
    case 'DAILY': {
      const next = new Date(from)
      next.setDate(next.getDate() + interval)
      return next
    }

    case 'WEEKLY': {
      if (rule.byweekday && rule.byweekday.length > 0) {
        const sortedDays = [...new Set(rule.byweekday)].sort((a, b) => a - b)
        const fromMonday = startOfWeek(from)
        // 搜索上限：interval 周 + 一周余量
        const limit = 7 * interval + 7
        for (let i = 1; i <= limit; i++) {
          const candidate = new Date(from)
          candidate.setDate(candidate.getDate() + i)
          const day = candidate.getDay() // 0=周日..6=周六
          if (sortedDays.includes(day)) {
            const candMonday = startOfWeek(candidate)
            const weekDiff = Math.round((candMonday.getTime() - fromMonday.getTime()) / MS_PER_WEEK)
            if (weekDiff >= 0 && weekDiff % interval === 0) {
              return candidate
            }
          }
        }
        return null
      } else {
        const next = new Date(from)
        next.setDate(next.getDate() + 7 * interval)
        return next
      }
    }

    case 'MONTHLY': {
      const next = new Date(from)
      const originalDay = from.getDate()
      next.setMonth(next.getMonth() + interval)
      // 处理溢出（如 1月31日 + 1月 → 3月3日），回退到目标月最后一天
      if (next.getDate() !== originalDay) {
        next.setDate(0) // setDate(0) = 上月最后一天（即目标月）
      }
      return next
    }

    case 'YEARLY': {
      const next = new Date(from)
      const originalMonth = from.getMonth()
      const originalDay = from.getDate()
      next.setFullYear(next.getFullYear() + interval)
      // 处理闰年 2月29日溢出
      if (next.getMonth() !== originalMonth || next.getDate() !== originalDay) {
        next.setMonth(originalMonth + 1, 0) // 目标月最后一天
      }
      return next
    }

    default:
      return null
  }
}

/**
 * 根据规则计算从 from 之后的下一个出现日期。
 * 返回 null 表示规则已到期（endDate/count 到达）或无法计算。
 */
export function getNextOccurrence(rule: RepeatRule, from: Date): Date | null {
  // count 已耗尽：count<=0 表示不再生成新出现
  if (rule.count !== undefined && rule.count <= 0) return null

  const next = computeNextDate(rule, from)
  if (!next) return null

  // endDate 到期检查
  if (rule.endDate) {
    const endDate = new Date(rule.endDate)
    if (!isNaN(endDate.getTime()) && next > endDate) return null
  }

  return next
}

/**
 * 返回重复规则的中文描述，如 "每周一、三、五"、"每2周"、"每天"。
 * rule 为 null 时返回空字符串。
 */
export function getRepeatSummary(rule: RepeatRule | null): string {
  if (!rule) return ''

  switch (rule.freq) {
    case 'DAILY':
      return rule.interval > 1 ? `每${rule.interval}天` : '每天'

    case 'WEEKLY': {
      if (rule.interval > 1) {
        return rule.byweekday && rule.byweekday.length > 0
          ? `每${rule.interval}周的周${formatWeekdays(rule.byweekday)}`
          : `每${rule.interval}周`
      }
      if (rule.byweekday && rule.byweekday.length > 0) {
        return `每周${formatWeekdays(rule.byweekday)}`
      }
      return '每周'
    }

    case 'MONTHLY':
      return rule.interval > 1 ? `每${rule.interval}个月` : '每月'

    case 'YEARLY':
      return rule.interval > 1 ? `每${rule.interval}年` : '每年'

    default:
      return ''
  }
}

/**
 * 将星期数字数组格式化为中文，如 [1,3,5] → "一、三、五"。
 * 输出已含"周"前缀的星期部分（不含"周"字本身，由调用方拼接）。
 */
function formatWeekdays(days: number[]): string {
  const sorted = [...new Set(days)].sort((a, b) => a - b)
  return sorted.map((d) => WEEKDAY_NAMES[d]).join('、')
}
